api: test existence of pid's

rerowep opened the pull request:
* Adds functionality to test pids for iexistence.
* Test during acq_accounts creation and update existence of pids.
* Test during acq_invoices cration and update existence of pids.
* Test during acq_order_lines creation and update existence of pids.
* Test during acq_orders creation and update existence of pids.
* Test during holdings creation and update existence of pids.
* Test during items creation and update existence of pids.
* Test during loans creation and update existence of pids.
* Test during patron_types creation und update existence of pids.

Co-Authored-by: Peter Weber <peter.weber@rero.ch>

## Why are you opening this PR?

- https://tree.taiga.io/project/rero21-reroils/task/1251

## How to test?


## Code review check list

- [ ] Commit message template compliance.
- [ ] Commit message without typos.
- [ ] File names.
- [ ] Functions names.
- [ ] Functions docstrings.
- [ ] Unnecessary commited files?
- [ ] Extracted translations?

BadrAly commented on rero_ils/modules/acq_accounts/api.py:
```diff
@@ -56,6 +56,17 @@ class AcqAccount(IlsRecord):
+        },
+        'not_required': {
+            'org': 'organisation',
+            'patr': 'patron'
```
I do not think there is a link to the patron record here.

rerowep replied:
You are right.

BadrAly commented on rero_ils/modules/acq_order_lines/api.py:
```diff
@@ -81,6 +82,17 @@ class AcqOrderLine(IlsRecord):
+        },
+        'not_required': {
+            'org': 'organisation',
+            'acor': 'acq_order'
```
The link to `acq_order` is required

rerowep replied:
ok

BadrAly commented on rero_ils/modules/acq_orders/api.py:
```diff
@@ -66,22 +76,23 @@ def create(cls, data, id_=None, delete_pid=False,
+        library = data.get('library', {})
+        library_pid = library.get('pid') or \
+            library.get('$ref').split('libraries/')[1]
+        org_ref = {'$ref': '{base_url}/api/{doc_type}/{pid}'.format(
```
This code is repeated in almost every resource. a separate function in the `utils`?

rerowep replied:
ok

BadrAly commented on rero_ils/modules/circ_policies/api.py:
```diff
@@ -58,6 +58,18 @@ class CircPolicy(IlsRecord):
+    pids_exist_test = {
+        'required': {
+            'org': 'organisation',
+            'budg': 'budget'
```
`budget` has no links with the `cipo`

rerowep replied:
TODO:  are in list of settings
```
         'ptty': 'patron_type',
         'itty': 'item_type'
```

BadrAly commented on rero_ils/modules/items/api.py:
```diff
@@ -215,6 +215,18 @@ class Item(IlsRecord):
+        },
+        'not_required': {
+            'org': 'organisation',
+            'hold': 'holding'
```
`holding ` is required

rerowep replied:
ok

BadrAly commented on rero_ils/modules/notifications/api.py:
```diff
@@ -69,6 +69,13 @@ class Notification(IlsRecord):
+    #         'loan': 'loan',
+    #     },
+    #     'raise_on_erro': True
+    # }
```
For now we generate notifications for loans only.

rerowep replied:
ok, deleted

BadrAly commented on rero_ils/modules/patron_transaction_events/api.py:
```diff
@@ -57,6 +57,16 @@ class PatronTransactionEvent(IlsRecord):
+            'lib': 'library'
+        },
+        'not_required': {
+            'ptrn': 'patron',
```
`patron` is required

rerowep replied:
ok

BadrAly commented on rero_ils/modules/patron_transaction_events/api.py:
```diff
@@ -57,6 +57,16 @@ class PatronTransactionEvent(IlsRecord):
+        },
+        'not_required': {
+            'ptrn': 'patron',
+            'pttr': 'patron_transaction'
```
`pttr` is required

rerowep replied:
ok

BadrAly commented on rero_ils/modules/acq_accounts/api.py:
```diff
@@ -56,6 +56,17 @@ class AcqAccount(IlsRecord):
     minter = acq_account_id_minter
     fetcher = acq_account_id_fetcher
     provider = AcqAccountProvider
+    pids_exist_test = {
```
`pids_exist_test` can be renamed to something similar to `pids_exist_check`.. to avoid confusion with the real `tests` units testing.

rerowep replied:
changed to: `pids_exist_check`

rerowep commented:
> Missing still the patron_types, ebooks, invoices.

* ebooks: are covert by documents.
* patron_types: Added `pattern` to JSON schema and configured required `org`
* acq_invoices: Added config: 
´´´
{
            'required': {
                'lib': 'library',
                'vndr': 'vendor'
            },
            'not_required': {
                'org': 'organisation'
            },
            'raise_on_erro': True
    }
```
